_posts/2020-07-05-gpg-card-with-google-vms.md

```diff
@@ -0,0 +1,110 @@
+
+`gcloud compute ssh` is just a tiny wrapper around `ssh` that does a couple of things on top of it:
+1. uses `~/.ssh/google_compute_engine(.pub)` SSH (public) key file path instead of standard
+   `~/.ssh/id_rsa(.pub)`,
+2. generates the key-pair automatically if it doesn't exist,
+3. updates target VM metadata (its `~/.ssh/authorized_keys`) with the public SSH key before
+   connecting (the convenient, and tricky, part).
+
+It should be possible to fool `gcloud compute ssh` by supplying the OpenPGP-backed public SSH key
+at the location it expects it, let's try:
+
+```bash
+$ ssh-add -L > ~/.ssh/google_compute_engine.pub
+$ gcloud compute ssh bencher
+WARNING: The private SSH key file for gcloud does not exist.
+WARNING: Your SSH key files are broken.
+private key (NOT FOUND) [/home/strohel/.ssh/google_compute_engine]
+public key  (OK)        [/home/strohel/.ssh/google_compute_engine.pub]
+We are going to overwrite all above files.
+Do you want to continue (y/N)?  N
+```
+
+Hmmm, `gcloud` was being a bit too smart here. We don't have the private key in file, because that's
+the entire point of confining it into a separate hardware device. Let's try to cheat a bit more:
+
+```bash
+$ touch ~/.ssh/google_compute_engine
+$ gcloud compute ssh bencher
+WARNING: The private SSH key file for gcloud is empty.
+(...)
+```
+
+Still no avail. Let's outsmart `gcloud` completely:
+
+```bash
+$ echo "fake content, actual private key on OpenPGP HW token" > ~/.ssh/google_compute_engine
+$ gcloud compute ssh bencher
+(...)
+Last login: ...
+```
+
+Yay. That's it, we need to write the public key in SSH format into `google_compute_engine.pub`, and
+some random content into `google_compute_engine`.
+
+## Notes
+
+1. I have tested this successfully with [YubiKey 4](https://support.yubico.com/support/solutions/articles/15000006486-yubikey-4)
+   and [ZeitControl "BasicCard" SmartCard (no longer) provided by FSFE](https://fsfe.org/news/2017/news-20171116-01.en.html).
+1. If you try to SSH without your OpenPGP token available, the error will look like:
+   ```bash
+   $ gcloud compute ssh bencher
+   Load key "/home/strohel/.ssh/google_compute_engine": invalid format
+   (...)
+   ```
+1. [Some folks say they need to write the authentication keygrip into `~/.gnupg/sshcontrol` file.](
+   https://stackoverflow.com/a/48922829/4345715) For me this works without touching the `sshcontrol`
+   file (it contains only comments). I suspect that it is used only for file-based GPG-as-SSH keys,
+   and not relevant for GPG-as-SSH keys backed by HW tokens.
+1. To resolve some problems with card "PIN Entry" dialog not showing up sometimes, I have this
+   snippet in my `~/.bashrc`:
+   ```bash
+   # Needed to make gpg-agent as ssh-agent pop-up pinentry correctly
+   # https://unix.stackexchange.com/a/218401
+   echo UPDATESTARTUPTTY | gpg-connect-agent >/dev/null
+   ```
+1. Sometimes, during SSH operations like git pushes, I get `Agent refused operation` errors from
```

Comment:
nitpik: `git pushes` sounds odd, but might be valid english -> I would do `operations like git push`

Reply:
Will do, backticks to the rescue:

during SSH operations like `git push`, ...